### ISSUE TITLE
Unable to find a matching function overload

### ISSUE THREAD

**Issue by schperplata** [microsoft/pylance-release#549]
Issue Type: <b>Question</b>
Is there anything obviously wrong with my type hints that VSC complains about: `No overloads for "colorize(data, 'red')" match parameters`?
```python
@overload
def colorize(data: str, color: str) -> str: ...
@overload
def colorize(data: List[str], color: str) -> List[str]: ...

def colorize(data: Union[List[str], str], color: str) -> Union[List[str], str]:
    if isinstance(data, list):
        if len(data):
            data[0] = f"<div style=\"color: {color};\">{data[0]}"
            data[-1] = f"{data[-1]}</div>"
    else:
        data = f"<span style=\"color: {color};\">{data}</span>"

    return data


@overload
def inRedColor(data: str) -> str: ...
@overload
def inRedColor(data: List[str]) -> List[str]: ...


def inRedColor(data: Union[List[str], str]) -> Union[List[str], str]:
    return colorize(data, 'red')
```
Error:
![image](https://user-images.githubusercontent.com/8736111/97848421-57c29c80-1cf1-11eb-9acc-eb1f58fc61eb.png)

Extension version: 2020.10.3
VS Code version: Code 1.50.1 (d2e414d9e4239a252d1ab117bd7067f125afd80a, 2020-10-13T15:06:15.712Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz (8 x 2712)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.92GB (5.83GB free)|
|Process Argv|W:\AsystLib\isys\python\VSCodeDevWorkspace\_refactor.code-workspace --crash-reporter-id 919e0ed9-46f3-4ad4-9e28-f35755bf8266|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by erictraut** [comment]
This is related to this issue: https://github.com/microsoft/pyright/issues/1119

It speaks to a subtle difference between the way pyright/pylance handle overloads versus mypy. We've talked about possibly switching to a behavior that is consistent with mypy, but it's a big code change and has potential significant performance impact.

**Comment by erictraut** [comment]
You can easily work around the issue by adding a third overload:
```python
@overload
def colorize(data: Union[List[str], str], color: str) -> Union[List[str], str]: ...
```

**Comment by erictraut** [comment]
Reopening this issue. We've received a sufficient number of bug reports about the overload matching behavior that I decided to change the behavior to more closely match mypy's. This eliminates the need for workarounds like the one I posted above. This change will be in the next version of pylance.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.4.3, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202143-29-april-2021
